File: Cheapest Flights Within K Stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        

        // using concepts of dijkstra's algorithm ;
        vector< pair<int,int>> adj[n] ;
        for(auto it : flights) {
            adj[it[0]].push_back({it[1] , it[2]}) ;
        }

        // { stop , {node , dist}} ;
        queue< pair<int , pair<int,int>>> q ;
        q.push({0 , {src , 0} }) ;

        vector<int> dist(n , 1e9) ;
        dist[src] = 0 ;
 
        while(!q.empty()) {
            auto it = q.front() ;
            q.pop() ;

            int stop = it.first ;
            int node = it.second.first ;
            int cost = it.second.second ;

            if( stop > k ) continue ;

            for(auto it : adj[node]) {
                int adjNode = it.first ;
                int edgeWeight = it.second ;

                if(cost + edgeWeight < dist[adjNode] && stop <= k ) {
                    dist[adjNode] = cost + edgeWeight ;
                    q.push({stop + 1 , {adjNode , cost+edgeWeight }}) ;
                }
            }
        }
        if(dist[dst] == 1e9) return -1 ;
        return dist[dst] ;
    }
};

File: Cheapest Flights Within K Stops.java
class Solution {
	public int findCheapestPrice(int n, int[][] flights, int src, int dst, int K) {

		/**
		src --> map(dest, cost);
		*/

		Map<Integer, Map<Integer, Integer>> map = new HashMap<>();
		for (int i = 0; i < flights.length; i++) {
			if (!map.containsKey(flights[i][0])) {
				map.put(flights[i][0], new HashMap<>());
			}
			map.get(flights[i][0]).put(flights[i][1], flights[i][2]);
		}

		Queue<int[]> q = new PriorityQueue<>((a, b) -> 
											 (Integer.compare(a[0], b[0])));
		int[] init = new int[]{0, src, K + 1};
		q.offer(init);

		while (!q.isEmpty()) {
			int[] cur = q.poll();
			int price = cur[0];
			int city = cur[1];
			int stops = cur[2];
			if (city == dst) return price;
			if (stops > 0) {
				Map<Integer, Integer> map2 = 
					map.getOrDefault(city, new HashMap());
				for (int a : map2.keySet()) {
					q.add(new int[] {price + map2.get(a), a, stops - 1});
				}
			}
		}
		return -1;
	}
}
